### ISSUE TITLE
Sockets.js afterDisconnect being called after Sails.lower

### ISSUE THREAD

**Issue by wedneyyuri** [balderdashy/sails#3332]
Hi Guys, I'm trying to do some database manipulation using the `afterDisconnect` handler in [sockets.js](http://sailsjs.org/documentation/anatomy/my-app/config/sockets-js):
```
  afterDisconnect: function(session, socket, cb) {
    var socketID = sails.sockets.id(socket);
    if (!socketID) {
      return cb(new Error('socketID is null!'));
    }

    User.removeSocket(socketID)
    .then(function() {
      cb();
    }, cb);
  },
```
but, after calling `sails.lower(done);` in [bootstrap.test.js](http://sailsjs.org/documentation/concepts/testing#/bootstrap-test-js) I'm getting this error:
```
  1)  "after all" hook:
     Uncaught TypeError: Cannot read property 'update' of undefined
      at Object.module.exports.adapter.update (C:\Users\wedne_000\Documents\Estudo\rbk-server\node_modules\sails-memory\lib\adapter.js:125:27)
      at module.exports.update (C:\Users\wedne_000\Documents\Estudo\rbk-server\node_modules\sails\node_modules\waterline\lib\waterline\adapter\dql.js:241:13)
      at bound.updateRecords (C:\Users\wedne_000\Documents\Estudo\rbk-server\node_modules\sails\node_modules\waterline\lib\waterline\query\dql\update.js:201:16)
      at C:\Users\wedne_000\Documents\Estudo\rbk-server\node_modules\sails\node_modules\waterline\lib\waterline\query\dql\update.js:52:21
```

So, after some debugging, I found a private property (`_exiting`) in `node_modules\sails\lib\app\lower.js`. So, my current workaround is:
```
  afterDisconnect: function(session, socket, cb) {
    // see node_modules\sails\lib\app\lower.js:32
    // if sails is exiting we can't do database operations
    if (sails._exiting === true) {
      return cb();
    }

    var socketID = sails.sockets.id(socket);
    if (!socketID) {
      return cb(new Error('socketID is null!'));
    }

    User.removeSocket(socketID)
    .then(function() {
      cb();
    }, cb);
  },
```
* Is this the expected behavior?
* Is there any better way to do this?

**Comment by sailsbot** [comment]
Thanks for posting, @wedneyyuri.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by GeorgeSapkin** [comment]
The issue is still there. It shouldn't be closed.

**Comment by wedneyyuri** [comment]
+1. It's a undocumented behavior.
